### PR TITLE
RES-2032: Gradient Metrics Mixin

### PR DESCRIPTION
**Pull request by alexcuozzo-1**
The base code is there and I have gotten some initial plots, but I am not sure if I am storing the figures in the correct way in the results dict. Also, as Subutai recommended I'm going to be updating the unit test to double check that the gradients are being stored/used correctly.

### CODE REVIEW

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+            - plot_freq: (optional) how often to create the plot, measured in training
+                         iterations; defaults to 1
+            - metrics: a list of metrics options from ["cosine", "dot",
+            "pearson"]; defaults to ["cosine",]
+            - gradient_transformation: (optional) can be one of "real", "sign",
+            "mask". "real" corresponds to the real values of the gradients,
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:54]: Please fix hanging indents for 'metrics' and 'gradient_transformation'. Do the same for 'max_samples_to_track' (below).

Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/gradient_metrics.py:54]: In fact, you started the new line too early.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            - include_names: (optional) a list of module names to track e.g.
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+            - plot_freq: (optional) how often to create the plot, measured in training
+                         iterations; defaults to 1
+            - metrics: a list of metrics options from ["cosine", "dot",
+            "pearson"]; defaults to ["cosine",]
+            - gradient_transformation: (optional) can be one of "real", "sign",
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:53]: This should actually be 'gradient_values', right? That's what the example config shows, and also what you search for in `process_gradient_metrics_args`.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            if self.gradient_values == "sign":
+                gradients = torch.sign(gradients)
+            elif self.gradient_values == "mask":
+                gradients = torch.abs(torch.sign(gradients))
+            for metric in self.gradient_metrics:
+                if metric == "cosine":
+                    stats = [
+                        torch.cosine_similarity(x, y, dim=0)
+                        for x in gradients
+                        for y in gradients
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:174]: I always thought list comprehensions with tensors are "unpythonic", but I'm not too concerned about it (since the code still works). I don't have a strong opinion on this issue, but I'm curious to know what @mvacaporale thinks.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    def plot_gradient_metric_heatmaps(self, gradient_metrics_stats):
+        order_by_class = torch.argsort(self.gradient_metric_targets)
+        stats_and_figures = []
+        for (name, module, metric, stats) in gradient_metrics_stats:
+            stats = stats[order_by_class, :][:, order_by_class]
+            plt.cla()
+            fig, ax = plt.subplots()
+            max_val = np.abs(stats).max()
+            # change to red/blue colormap
+            # after finishing, screenshot a demo and put into PR
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:199]: These comments should be removed.

File: nupic/research/frameworks/vernon/mixins/__init__.py [hunk 5]
```diff
@@ -55,3 +55,4 @@
 from .update_boost_strength import UpdateBoostStrength
 from .update_dendrite_boost_strength import UpdateDendriteBoostStrength
 from .vary_batch_size import VaryBatchSize
+from .gradient_metrics import GradientMetrics
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/__init__.py:58]: (very minor comment) I just sorted this init file alphabetically in my last merge, so I ask that you keep it that way. You can simply run the `isort` command on the file from the command line and it will do it automatically for you.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+        # Log the names of the modules being tracked
+        tracked_names = pformat(list(named_modules.keys()))
+        self.logger.info(f"Tracking gradients for modules: {tracked_names}")
+
+        # The targets will be collected in `self.error_loss` in a 1:1 fashion
+        # to the tensors being collected by the hooks.
+        self.gradient_metric_targets = torch.tensor([]).long()
+
+    def process_gradient_metrics_args(self, gradient_metric_args):
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:100]: Since this method doesn't use `self`, I recommend making it a separate function in this same module.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        if iteration % self.gradient_metrics_plot_freq == 0:
+            gradient_stats = self.gradient_metric_hooks.get_statistics()
+            gradient_metrics_stats = self.calculate_gradient_metrics_stats(
+                gradient_stats
+            )
+            gradient_metric_heatmaps = self.plot_gradient_metric_heatmaps(
+                gradient_metrics_stats
+            )
+            for (name, _, metric, _, figure) in gradient_metric_heatmaps:
+                results.update({f"{name}/{metric}": figure})
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:159]: A small note on naming: The naming helps group plots together on visualization dashboards (e.g. wandb or tensorboard). If you start the naming with `{name}/`, then you'll group the plots according to the module. If you start the naming with `{metric}/`, then you'll group them according to the metric. This may be more preferable, as you may want to compare the same metric accrose the different modules. You'll get the same plots, but one grouping may be easier to make the comparisons you're interested in. Up to you.

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        return results
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/gradient_metrics.py:160]: Have you tested this with Ray? It usually doesn't respond well when you attempt to log a plot. It inevitably tries to deepcopy the plot, which can't be done since that's not implemented by matplotlib.

Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/gradient_metrics.py:160]: One thing you can do is wrap the plot so it's in a loggable form. Say you accept an additional argument to your mixin called `prep_plots_for_wandb` (with a default False so we're wandb agnostic), then you can have logic like
```
from nupic.research.frameworks.wandb import ray_wandb

...
if self.prep_plots_for_wandb:
    figure = ray_wandb.prep_plots_for_wandb(figure)
results.update({f"{metric}/{name}": figure})
```

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py (hunk 3 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/gradient_metrics.py:174]: I do believe list comprehensions have their place. Here they work well, in my opinion. I believe they're problematic when they make the code difficult to read. As far as list-comprehensions with tensors specifically, I'm not sure. Can you share where you saw or read that? Just curious for myself.

**Review by mvacaporale — APPROVED** [review]
Only minor comments. Otherwise, looks good to go. I'm approving barring making sure that the mixin can indeed run with Ray without throwing an error (see comment below).

File: nupic/research/frameworks/vernon/mixins/gradient_metrics.py (hunk 3 above)
Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/gradient_metrics.py:174]: @mrcslws once mentioned this to me, and linked to [this stack overflow article](https://stackoverflow.com/questions/5753597/is-it-pythonic-to-use-list-comprehensions-for-just-side-effects) -- but now I don't see anything about tensors mentioned there.

**Review by karangrewal — APPROVED** [review]
